File: src/_300.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int lengthOfLIS(vector<int> &nums)
    {
        int len = nums.size();
        vector<int> dp(len + 1, 0);
        int ans = 0;
        dp[len] = 0;
        int t, j, m;
        for (int i = len - 1; i >= 0; --i)
        {
            t = nums[i];
            for (j = i + 1, m = 0; j < len; ++j)
            {
                if (nums[j] > t)
                {
                    m = max(dp[j],m);
                }
            }
            dp[i] = m + 1;
            ans = max(ans, m+1);
        }
        return ans;
    }
};

int main()
{
    Solution test;
    vector<int> a = {0, 1, 0, 3, 2, 3};
    cout << test.lengthOfLIS(a) << endl;
    return 0;
}

File: 300.py
from typing import List
from bisect import bisect_left

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [1] * n
        for i in range(1, n):
            for j in range(i-1, -1, -1):
                if dp[i] > j + 1:
                    break
                if nums[i] > nums[j]:
                    dp[i] = max(dp[i], dp[j] + 1)
        return max(dp)
    def lengthOfLIS(self, nums: List[int]) -> int:
        g = []
        for x in nums:
            j = bisect_left(g, x)
            if j == len(g):
                g.append(x)
            else:
                g[j] = x
        return len(g)